### PR TITLE
moving plugin template to React and EUI

### PR DESCRIPTION
**Pull request by bmcconaghy**
updating Kibana plugin template to use React and EUI.

### CODE REVIEW

**Comment by kimjoar** [comment]
@bmcconaghy We recently moved the plugin generator to https://github.com/elastic/kibana/tree/master/packages/kbn-plugin-generator, for all new plugins (6.3 and greater). We definitely want this change there, but not sure about applying it here. Maybe move this PR there first, so we can merge it there, then potentially "backport" it here (aka reopen this PR)?

File: template/public/app.js [hunk 1]
```diff
@@ -1,34 +1,48 @@
...
 
-import 'ui/autoload/styles';
-import './less/main.less';
-import template from './templates/index.html';
+import "ui/autoload/styles";
+import "./less/main.less";
+import template from "./templates/index.html";
+import { Main } from "./components/main";
 
+const app = uiModules.get("apps/<%= name %>");
```
Comment (cjcenizal) [inline template/public/app.js:13]: This is partially a matter of personal preference, but can we add whitespace to break the code up into clearer chunks? I think I'd find this code easier to scan if we add newlines on line 14, 17, 35, and 36.

File: template/public/app.js [hunk 2]
```diff
@@ -1,34 +1,48 @@
...
+app.config(stateManagementConfigProvider => {
+  stateManagementConfigProvider.disable();
+});
+const controller = ($scope, $route, $interval) => {
```
Comment (cjcenizal) [inline template/public/app.js:17]: Likewise, I'd be able to scan this function more easily with newlines on lines 20, 25, 26, and 30.

File: template/public/app.js [hunk 3]
```diff
@@ -1,34 +1,48 @@
...
+  const elem = document.getElementById("pluginReactRoot");
+  let currentTime = moment($route.current.locals.currentTime);
+  const props = {
+    title: "<%= this.startCase(name) %>",
+    description: "<%= description %>",
+    currentTime: currentTime.format("HH:mm:ss")
+  };
+  render(<Main {...props} />, elem);
```
Comment (cjcenizal) [inline template/public/app.js:25]: Can we add a comment here:

```jsx
// Mount the React component tree on the DOM node.
render(<Main {...props} />, elem);
```

File: template/public/app.js [hunk 4]
```diff
@@ -1,34 +1,48 @@
...
+  const stop = $interval(function() {
```
Comment (cjcenizal) [inline template/public/app.js:26]: And another one here?

```jsx
// Every time the props change, we need to re-mount the React component tree so that it updates the DOM.
currentTime = currentTime.add(1, "second");
render(<Main {...props} currentTime={currentTime.format("HH:mm:ss")} />, elem);
```

File: template/public/app.js [hunk 5]
```diff
@@ -1,34 +1,48 @@
...
+    currentTime = currentTime.add(1, "second");
+    render(<Main {...props} currentTime={currentTime.format("HH:mm:ss")} />, elem);
+  }, 1000);
+  $scope.$watch("$destroy", () => {
+    $interval.cancel(stop);
+    elem && unmountComponentAtNode(elem);
```
Comment (cjcenizal) [inline template/public/app.js:32]: And one last one here?

```js
// To clean up after ourselves, we need to tell React to unmount from the DOM node.
elem && unmountComponentAtNode(elem);
```

File: template/public/components/main/main.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+export const Main = ({ title, currentTime }) => {
+  return (
+    <EuiPage>
+      <EuiPageHeader>
+        <EuiTitle size="l">
+          <h1>{title} Hello World!</h1>
+        </EuiTitle>
+      </EuiPageHeader>
+      <EuiPageBody>
```
Comment (cjcenizal) [inline template/public/components/main/main.js:21]: I've gotten in the habit of adding a newline between large adjacent elements in JSX, which helps me scan and form an idea of the resulting DOM structure. Applied here, this would add newlines to line 21 and 28. Something to consider.

**Review by cjcenizal — APPROVED** [review]
Looks great man! I just had some suggestions to improve readability and add some comments.

**Comment by bmcconaghy** [comment]
@cjcenizal made these changes in the corresponding PR for Kibana.  This repo is deprecated apparently.

**Comment by kimjoar** [comment]
(The deprecation is mentioned in the first line in the readme. We haven’t archived the github repo as it’s still used for Kibana versions prior to 6.3)

File: template/public/app.js [hunk 7]
```diff
@@ -1,34 +1,48 @@
...
+import "./less/main.less";
+import template from "./templates/index.html";
+import { Main } from "./components/main";
 
+const app = uiModules.get("apps/<%= name %>");
+app.config(stateManagementConfigProvider => {
+  stateManagementConfigProvider.disable();
+});
+const controller = ($scope, $route, $interval) => {
+  const elem = document.getElementById("pluginReactRoot");
```
Comment (w33ble) [inline template/public/app.js:18]: I see what you're doing here with the template and such. This isn't how any of the other React apps work, as far as I know. There's two methods I've seen in use so far, a react directive, and using Kibana's root controller. Here's how the latter works:

```js
import React from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import chrome from 'ui/chrome';
import { Main } from './components/main/main';

function RootController($scope, $element) {
  const domNode = $element[0];

  // render react to DOM
  render(<Main />, domNode);

  // unmount react on controller destroy
  $scope.$on('$destroy', () => {
    unmountComponentAtNode(domNode);
  });
}

chrome.setRootController('myplugin', RootController);
```

You can drop this into the top-level `app.js` file, and then kick off the actual application in `components/main/main`. Paths and naming can all be changed here, but this separates the angular bootstrapping from the react app itself. This also means we can render the react app without using the angular router, and can avoid enabling that entirely.

@spalger any additional input here?

File: template/public/app.js (hunk 3 above)
Comment (w33ble) [reply template/public/app.js:25]: Do we need this anymore? Can we just move all this into the react part of the app, and keep the angular bootstrapping really minimal? None of the props here require angular...

Removing the moment dependency from the template would also be great.

File: template/public/app.js (hunk 7 above)
Comment (w33ble) [reply template/public/app.js:18]: Also, if we're in favor of not using the angular router, we should also disable the location provider:

```js
app.config($locationProvider => {
  $locationProvider.html5Mode({
    enabled: false,
    requireBase: false,
    rewriteLinks: false,
  });
});
```

Otherwise it'll rewrite links by default and break whatever other router is being used.

**Review by w33ble — COMMENTED** [review]
tl;dr I think we should keep the angular surface as small as possible

**Comment by spalger** [comment]
I like your feedback @w33ble, but I would really to reiterate @kjbekkelund and suggest that this be reopened and reviewed in the Kibana repo and backported from there if desired.

**Comment by w33ble** [comment]
@spalger near as I can tell, this is https://github.com/elastic/kibana/pull/16937 applied to the standalone tool here. FWIW, I'm in favor of closing this PR until we settle on the changes there. Concurrent PR's are just going to muddy the waters (I didn't realize that was happening until after I left comments here).
